---
title: "Analysis of variants potentially associated with malaria"
author: 
  name: "Yvon Mbouamboua & Jacques van Helden"
affiliation: "Theories and Approach Genomic Complexity (TAGC)"
email: fridyvon@gmail.com
#date: 'Last update: `r Sys.Date()`'
date: Last update:`r format(Sys.time())`
output:
  html_document:
    code_folding: hide
    fig_caption: yes
    highlight: zenburn
    self_contained: yes
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: yes
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 3
  word_document: default
---

```{r libraries, include=FALSE,  eval=TRUE, echo=FALSE, warning=FALSE}
## Global configuration to generate the documents from this Rmd file
libraries.to.install <- c(
  "knitr",
  "readtext",
  "RCurl", 
  "ggplot2",
  "gridExtra", 
  "cowplot", 
  "dplyr",
  "tidyr",
  "VennDiagram", 
  "writexl", 
  "stringr", 
  "gdata", 
  "PythonInR", 
  "reticulate", 
  "haploR", 
  "RMySQL", 
  "GEOquery", 
  "gProfileR", 
  "lattice", 
  "qqman", 
  "readr", 
  "httr", 
  "jsonlite", 
  "xml2")

message("Loading libraries")
for (lib in libraries.to.install) {
  if (require(lib, character.only = TRUE, quietly = TRUE)) {
    message("\tLoaded library\t", lib)
  } else {
    message("Installing library\t", lib)
    install.packages(lib, dependencies = TRUE)
  }
}
```


```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo=FALSE, 
  eval=TRUE, 
  cache=TRUE, 
  message=FALSE, 
  warning=FALSE, 
  comment = "",  
  fig.align= "center",
  fig.width = 7, 
  fig.height = 5,
  fig.path = "figures/")

## Parameters
parameters <- list(
  update.flowcharts = TRUE, # Update the flowcharts with graphviz dot
  flowchart.formats = c("pdf", "png"), # List of formats to generate
  flowchart.format = "png" # Format for insertion in the report
)


## IDs of the diseases of interesd
diseaseIDs <- c("Malaria" = "C0024530", 
                "Malaria susceptibility" = "C1970028",
                "Antepartum malaria" = "C0747820",
                "Mild malaria susceptibility" = "C1836721",
                "Cerebral malaria" = "C0024534",
                "Cerebral malaria susceptibility" = "C1855457",
                "Cerebral malaria resistance" = "C1969379")

## Summarize the genes per disease from disgenet
gene.summary <- data.frame(
  row.names = diseaseIDs,
  ID = diseaseIDs,
  name = names(diseaseIDs))

## Get source
source("R/get_genes_for_disease.R")
```



```{r configuration}
message("Defining working directory and subfolders")

## Main directory
dir.main <- '~/Google\ Drive/These_Yvon_2018/malaria_rSNPs'

# Result directory (export result tables)
dir.results <- file.path(dir.main, "results")
result.folders <- list(gene.disgenet = "gene.disgenet",
                       snp.disgenet = "snp.disgenet", 
                       gwas.2014 = "gwas.2014",
                       gwas.2018 = "gwas.2018",
                       remap = "remap",
                       roken = "roken",
                       tag.snps = "tag.snps",
                       haploreg = "haploreg",
                       sois = "sois",
                       soi.motifs = "soi.motifs")

dir.path <- c()
for (folder in result.folders) {
  dir.path[folder] <- file.path(dir.results, folder)
  dir.create(dir.path[folder], showWarnings = FALSE, recursive = TRUE)
}

message("\tWorking directory: ", getwd())

#setwd(dir.main)
```


# Goal

1) Identify the variants potentially associated with malaria.


# Bibliographic search

- Genes associated with malaria resistance.

- Look for rSNPs around these genes.

- Phenotypes studied:
  
  - Parasitemia
- Cerebral malaria
- Respiratory distress
- Severe anemia


# GWAS search

- There are annotations of some symptoms in the GWAS catalog.
- To cross-search data of genes / SNPs associated with malaria.


# Search criteria for SNPs

- Number of studies / number of different populations where this variant was detected.
- Functional analyzes of the mutant.


# Flow chart 

```{r generate_flowcharts, eval=TRUE}

message("Updating flowchart figures")
for (flowchart in c(
  "flowchart/gene.wise.approach",
  "flowchart/snp.wise.approach")) {
  
  for (format in parameters$flowchart.formats) {
    outfile <- paste(sep="", flowchart, ".", format)
    cmd <- paste(sep="", "dot -T", format, 
                 " ",  flowchart,".dot",
                 " -o ", outfile)
    #    message("\t", format, "\t", cmd)
    system(cmd)  
    message("\t", outfile)
  }
}

```

## Gene-wise approach

```{r fig.cap="Flow chart of the gene-wise approach. "}
knitr::include_graphics(paste(sep = "", "flowchart/gene.wise.approach.", parameters["flowchart.format"]))
```

## SNP-wise approach

```{r fig.cap="Flow chart of the snp-wise approach. "}
knitr::include_graphics(paste(sep = "", "flowchart/snp.wise.approach.", parameters["flowchart.format"]))
```



## Gene associated with malaria from DisGeNet

http://www.disgenet.org/web/DisGeNET/menu;jsessionid=1x6jf5ghyr1ga155fxkzfk2iji

```{r downloading_GDAs}

message("Downloading malaria-associated genes from disgenet")
## Download gene table from DisGeNet
url <- "http://www.disgenet.org/oql"

geneTables <- list()
genes.per.disease <- list()
for (d in 1:length(diseaseIDs)) {
  diseaseID <- diseaseIDs[[d]]
  diseaseName <- names(diseaseIDs)[[d]]
  
  gene.data <- GetGenesForDisease(diseaseID, diseaseName)
  
  ## Export the full result table
  gene.data.file <- paste(sep = "",
                          diseaseID, "_",
                          gsub(pattern = " +", replacement = "_", 
                               x = diseaseName),
                          ".tsv")
  
  gene.data.path <- file.path(dir.path["gene.disgenet"], gene.data.file)
  message("\tgene table\t", gene.data.path)

  write.table(x = gene.data,
              file = gene.data.path,
              quote = FALSE,
              sep = "\t",
              row.names = FALSE,
              col.names = TRUE)
}

# venn.file <- file.path(dir.results, "genes.per.disease.Venn.tiff")
# venn.diagram(x = genes.per.disease, filename = venn.file) #, height = 7, width = 8, resolution = 72)
# venn.plot <- venn.diagram(x = genes.per.disease, filename = venn.file) #, height = 7, width = 8, resolution = 72)

```



## SNPs associated with malaria from DisGeNet

```{r downloading VDAs}
message("Downloading malaria-associated SNPs from disgenet")
## Download SNP table from DisGeNet
url <- "http://www.disgenet.org/oql"

snpTables <- list()
snp.per.disease <- list()
for (s in 1:length(diseaseIDs)) {
  diseaseID <- diseaseIDs[[s]]
  diseaseName <- names(diseaseIDs)[[s]]
  
  snp.data <- GetSnpsForDisease(diseaseID, diseaseName)
  
  ## Export the full result table
  snp.data.file <- paste(sep = "",
                          diseaseID, "_",
                          gsub(pattern = " +", replacement = "_", 
                               x = diseaseName),
                          ".tsv")
  
  snp.data.path <- file.path(dir.path["snp.disgenet"], snp.data.file)
  message("\tgene table\t", snp.data.path)

  write.table(x = snp.data,
              file = snp.data.path,
              quote = FALSE,
              sep = "\t",
              row.names = FALSE,
              col.names = TRUE)
  
}
```


```{r}
print(ggplot(snp.data, aes(c2.most_severe_consequence	)) + geom_bar() + coord_flip())
```


## Collecte the significant SNPs associated with malaria from to GWAS 2014 and 2018

```{r gwas_2014_2018}
gwas2014 <- read_tsv("data/gwas/gwas2014.tsv")
gwas2018 <- read_tsv("data/gwas/gwas2018.tsv")

## Select a subset of fields for display in the report
snp.gwas.fields <- c("Gene", "rsID", "Phenotype") 
  
## Display a small subset of the table
gwas2014.table <- head(gwas2014[, snp.gwas.fields])

gwas2018.table <- head(gwas2018[, snp.gwas.fields])

   
## Combine datset from GWAS 2014 and 2018
gwas.2014.vs.2018 <- rbind(gwas2014.table, gwas2018.table)

kable(gwas.2014.vs.2018)

## Plot the interest genes
print(ggplot(gwas.2014.vs.2018, aes(Phenotype)) + geom_bar() + coord_flip())
print(ggplot(gwas.2014.vs.2018, aes(Gene)) + geom_bar() + coord_flip())
```


## Collect Genes/SNPs malaria association from to GWAS catalog

Download the file that contains all associations with added ontology
either by clicking 
[this link](http://www.ebi.ac.uk/gwas/api/search/downloads/alternative), 
or with the wget command:
   
```{bash wget_data, eval=FALSE}
echo downloding GWAS catalog
wget http://www.ebi.ac.uk/gwas/api/search/downloads/alternative
mv alternative gwas_catalog.tsv
mv gwas_catalog.tsv ${HOME}/Google\ Drive/These_Yvon_2018/malaria_rSNPs/data/gwas.catalog
```


### Reading the GWAS catalog

```{r read_data, warning=FALSE, eval=TRUE, comment=NA}
message("Reading GWAS catalog in tsv format")
gwas <- read_tsv("data/gwas.catalog/gwas_catalog.tsv",na = c("","NA","NR"))
gwas <- as.data.frame(gwas,stringsAsFactors = FALSE)
#print(colnames(gwas))
#print(dim(gwas))

## Select the Genes/SNPs associated with malaria

malaria.gwas <-
  gwas %>% 
  rename("disease_trait" = "DISEASE/TRAIT",
         "pvalue" = "P-VALUE") %>% 
  filter(disease_trait %in% c("Malaria",
                              "Severe malaria",
                              "Severe malaria (adjusted for sickle cell variant rs334)"))

cat("Number of table rows before filtering step:",nrow(malaria.gwas))

#View(malaria.gwas)

## Select a subset of fields for display in the report
  snp.fields <- c(
    "chr" = "CHR_ID",
    "gene" = "MAPPED_GENE",
    "snp" = "SNPS",
    "locus" = "REGION",
    "phenotype" = "disease_trait",
    "pvalue" = "pvalue",
    "PMID" = "PUBMEDID"
  ) 
     
  ## Display a small subset of the table
  gwas.table <- head(malaria.gwas[, snp.fields])
  names(gwas.table) <- names(snp.fields)
   kable(gwas.table)
   
## Print the gwas.data plot
print(ggplot(gwas.table, aes(gene)) + geom_bar() + coord_flip())
   
```


### Malaria GWAS phenotypes summary plot

```{R trait_summary_plot, fig.height=7.5, fig.width=7.5, eval=TRUE}

print(ggplot(malaria.gwas, aes(disease_trait)) +
        geom_bar() + 
        coord_flip() +
        theme_minimal() +
        theme(panel.grid.major.y = element_blank(),
        panel.grid.minor.y = element_blank()))
```


### Significant malaria association SNPs from to GWAS catalog

```{r}
signif.snps.catalog.gwas <-
malaria.gwas %>% 
  filter( as.numeric(pvalue) <= 5e-08)
#View(signif.snps)

kable(signif.snps.catalog.gwas[, c("CHR_ID", "MAPPED_GENE", "REGION" ,"SNPS" , "disease_trait", "pvalue", "PUBMEDID" )])
```


