---
title: "Analysis of variants potentially associated with malaria"
author: 
  name: "Yvon Mbouamboua & Jacques van Helden"
affiliation: "Theories and Approach Genomic Complexity (TAGC)"
email: fridyvon@gmail.com
#date: 'Last update: `r Sys.Date()`'
date: Last update:`r format(Sys.time())`
output:
  html_document:
    code_folding: hide
    fig_caption: yes
    highlight: zenburn
    self_contained: yes
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: yes
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 3
  word_document: default
---

```{r libraries, include=FALSE,  eval=TRUE, echo=FALSE, warning=FALSE}
## Global configuration to generate the documents from this Rmd file
libraries.to.install <- c(
  "knitr",
  "readtext",
  "RCurl", 
  "ggplot2",
  "gridExtra", 
  "cowplot", 
  "dplyr",
  "tidyr",
  "VennDiagram", 
  "writexl", 
  "stringr", 
  "gdata", 
  "PythonInR", 
  "reticulate", 
  "haploR", 
  "RMySQL", 
  "GEOquery", 
  "gProfileR", 
  "lattice", 
  "qqman", 
  "readr", 
  "httr", 
  "jsonlite", 
  "xml2")

message("Loading libraries")
for (lib in libraries.to.install) {
  if (require(lib, character.only = TRUE, quietly = TRUE)) {
    message("\tLoaded library\t", lib)
  } else {
    message("Installing library\t", lib)
    install.packages(lib, dependencies = TRUE)
  }
}
```


```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo=FALSE, 
  eval=TRUE, 
  cache=TRUE, 
  message=FALSE, 
  warning=FALSE, 
  comment = "",  
  fig.align= "center",
  fig.width = 7, 
  fig.height = 5,
  fig.path = "figures/")

## Parameters
parameters <- list(
  update.flowcharts = TRUE, # Update the flowcharts with graphviz dot
  flowchart.formats = c("pdf", "png"), # List of formats to generate
  flowchart.format = "png" # Format for insertion in the report
)


## IDs of the diseases of interesd
diseaseIDs <- c("Malaria" = "C0024530", 
                "Malaria susceptibility" = "C1970028",
                "Antepartum malaria" = "C0747820",
                "Mild malaria susceptibility" = "C1836721",
                "Cerebral malaria" = "C0024534",
                "Cerebral malaria susceptibility" = "C1855457",
                "Cerebral malaria resistance" = "C1969379")

## Summarize the genes per disease from disgenet
gene.summary <- data.frame(
  row.names = diseaseIDs,
  ID = diseaseIDs,
  name = names(diseaseIDs))

## Get source
source("R/get_genes_for_disease.R")
source("R/get_snps_for_disease.R")
```



```{r configuration}
message("Defining working directory and subfolders")

## Main directory
dir.main <- '~/Google\ Drive/These_Yvon_2018/malaria_rSNPs'

# Result directory (export result tables)
dir.results <- file.path(dir.main, "results")
result.folders <- list(gene.disgenet = "gene.disgenet",
                       snp.disgenet = "snp.disgenet", 
                       gwas.2014 = "gwas.2014",
                       gwas.2018 = "gwas.2018",
                       remap = "remap",
                       roken = "roken",
                       tagsnps = "tagsnps",
                       haploreg = "haploreg",
                       sois = "sois",
                       soi.motifs = "soi.motifs")

dir.path <- c()
for (folder in result.folders) {
  dir.path[folder] <- file.path(dir.results, folder)
  dir.create(dir.path[folder], showWarnings = FALSE, recursive = TRUE)
}

message("\tWorking directory: ", getwd())

#setwd(dir.main)
```


# Goal

1) Identify the variants potentially associated with malaria.


# Bibliographic search

- Genes associated with malaria resistance.

- Look for rSNPs around these genes.

- Phenotypes studied:
  
  - Parasitemia
- Cerebral malaria
- Respiratory distress
- Severe anemia


# GWAS search

- There are annotations of some symptoms in the GWAS catalog.
- To cross-search data of genes / SNPs associated with malaria.


# Search criteria for SNPs

- Number of studies / number of different populations where this variant was detected.
- Functional analyzes of the mutant.


# Flow chart 

```{r generate_flowcharts, eval=TRUE}

message("Updating flowchart figures")
for (flowchart in c(
  "flowchart/gene_wise_approach",
  "flowchart/snp_wise_approach")) {
  
  for (format in parameters$flowchart.formats) {
    outfile <- paste(sep="", flowchart, ".", format)
    cmd <- paste(sep="", "dot -T", format, 
                 " ",  flowchart,".dot",
                 " -o ", outfile)
    #    message("\t", format, "\t", cmd)
    system(cmd)  
    message("\t", outfile)
  }
}

```

## Gene-wise approach

```{r fig.cap="Flow chart of the gene-wise approach. "}
knitr::include_graphics(paste(sep = "", "flowchart/gene_wise_approach.", parameters["flowchart.format"]))
```

## SNP-wise approach

```{r fig.cap="Flow chart of the snp-wise approach. "}
knitr::include_graphics(paste(sep = "", "flowchart/snp_wise_approach.", parameters["flowchart.format"]))
```



## Gene associated with malaria from DisGeNet

http://www.disgenet.org/web/DisGeNET/menu;jsessionid=1x6jf5ghyr1ga155fxkzfk2iji

```{r downloading_GDAs}

message("Downloading malaria-associated genes from disgenet")
## Download gene table from DisGeNet
url <- "http://www.disgenet.org/oql"

geneTables <- list()
genes.per.disease <- list()
gene.data.all.diseases <- data.frame()
for (d in 1:length(diseaseIDs)) {
  diseaseID <- diseaseIDs[[d]]
  diseaseName <- names(diseaseIDs)[[d]]
  
  gene.data <- GetGenesForDisease(diseaseID, diseaseName)
  gene.data.all.diseases <- rbind(gene.data.all.diseases, gene.data)
  
  ## Export the full result table
  gene.data.file <- paste(sep = "",
                          diseaseID, "_",
                          gsub(pattern = " +", replacement = "_", 
                               x = diseaseName),
                          ".tsv")
  gene.data.path <- file.path(dir.path["gene.disgenet"], gene.data.file)
  message("\tExporting gene table\t", gene.data.path)
  write.table(x = gene.data,
            file = gene.data.path,
            quote = FALSE,
            sep = "\t",
            row.names = FALSE,
            col.names = TRUE)

}

# dim(gene.data.all.diseases)
# table(gene.data.all.diseases$c2.diseaseId)


## Export the full result table
gene.data.path <- file.path(dir.path["gene.disgenet"], "all_diseases.tsv")
message("\tExporting gene table\t", gene.data.path)
write.table(x = gene.data,
            file = gene.data.path,
            quote = FALSE,
            sep = "\t",
            row.names = FALSE,
            col.names = TRUE)


# venn.file <- file.path(dir.results, "genes.per.disease.Venn.tiff")
# venn.diagram(x = genes.per.disease, filename = venn.file) #, height = 7, width = 8, resolution = 72)
# venn.plot <- venn.diagram(x = genes.per.disease, filename = venn.file) #, height = 7, width = 8, resolution = 72)

```



## SNPs associated with malaria from DisGeNet

```{r downloading VDAs}
message("Downloading malaria-associated SNPs from disgenet")
## Download SNP table from DisGeNet
url <- "http://www.disgenet.org/oql"

snpTables <- list()
snp.per.disease <- list()
snp.data.all.diseases <- data.frame()
for (s in 1:length(diseaseIDs)) {
  diseaseID <- diseaseIDs[[s]]
  diseaseName <- names(diseaseIDs)[[s]]
  
  snp.data <- GetSnpsForDisease(diseaseID, diseaseName)
  snp.data.all.diseases <- rbind(snp.data.all.diseases, snp.data)
  
  ## Export the full result table
  snp.data.file <- paste(sep = "",
                          diseaseID, "_",
                          gsub(pattern = " +", replacement = "_", 
                               x = diseaseName),
                          ".tsv")
  
  snp.data.path <- file.path(dir.path["snp.disgenet"], snp.data.file)
  message("\tgene table\t", snp.data.path)

  write.table(x = snp.data,
              file = snp.data.path,
              quote = FALSE,
              sep = "\t",
              row.names = FALSE,
              col.names = TRUE)
  
}
```


```{r}
print(ggplot(snp.data, aes(c2.most_severe_consequence	)) + geom_bar() + coord_flip())
```




## Collecte the significant SNPs associated with malaria from to GWAS 2014 and 2018

```{r gwas_2014_2018}
gwas2014 <- read_tsv("data/gwas/mgen_2014_signif_SNPs.tsv")
gwas2018 <- read_tsv("data/gwas/ravhenall_2018_signif_SNPs.tsv")

## Select a subset of fields for display in the report
snp.gwas.fields <- c("Gene", "rsID", "Phenotype") 
  
## Display a small subset of the table
gwas2014.table <- head(gwas2014[, snp.gwas.fields])

gwas2018.table <- head(gwas2018[, snp.gwas.fields])

   
## Combine datset from GWAS 2014 and 2018
gwas.2014.vs.2018 <- rbind(gwas2014.table, gwas2018.table)
View(gwas.2014.vs.2018)
kable(gwas.2014.vs.2018)

## Plot the interest genes
print(ggplot(gwas.2014.vs.2018, aes(Phenotype)) + geom_bar() + coord_flip())
print(ggplot(gwas.2014.vs.2018, aes(Gene)) + geom_bar() + coord_flip())
```


## Collect Genes/SNPs malaria association from to GWAS catalog

Download the file that contains all associations with added ontology
either by clicking 
[this link](http://www.ebi.ac.uk/gwas/api/search/downloads/alternative), 
or with the wget command:
   

### Reading the GWAS catalog

```{r read_data, warning=FALSE, eval=TRUE, comment=NA}
message("Downloding GWAS catalog")
download.file(url='http://www.ebi.ac.uk/gwas/api/search/downloads/alternative', destfile='data/gwas_catalog/gwas_catalog.tsv', method='auto')

message("Reading GWAS catalog in tsv format")
gwas <- read_tsv("data/gwas_catalog/gwas_catalog.tsv",na = c("","NA","NR"))
gwas <- as.data.frame(gwas,stringsAsFactors = FALSE)
#print(colnames(gwas))
#print(dim(gwas))

## Select the Genes/SNPs associated with malaria

malaria_gwas <-
  gwas %>% 
  rename("diseaseTrait" = "DISEASE/TRAIT",
         "pValue" = "P-VALUE") %>% 
  filter(diseaseTrait %in% c("Malaria",
                              "Severe malaria",
                              "Severe malaria (adjusted for sickle cell variant rs334)"))

cat("Number of table rows before filtering step:",nrow(malaria_gwas))

#View(malaria.gwas)

## Select a subset of fields for display in the report
  snp.fields <- c(
    "chr" = "CHR_ID",
    "gene" = "MAPPED_GENE",
    "snp" = "SNPS",
    "locus" = "REGION",
    "phenotype" = "diseaseTrait",
    "pValue" = "pValue",
    "PMID" = "PUBMEDID"
  ) 
     
  ## Display a small subset of the table
  gwas_table <- head(malaria_gwas[, snp.fields])
  names(gwas_table) <- names(snp.fields)
   kable(gwas_table)
   
## Print the gwas.data plot
print(ggplot(gwas_table, aes(gene)) + geom_bar() + coord_flip())
   
```


### Malaria GWAS phenotypes summary plot

```{R trait_summary_plot, fig.height=7.5, fig.width=7.5, eval=TRUE}

print(ggplot(malaria_gwas, aes(diseaseTrait)) +
        geom_bar() + 
        coord_flip() +
        theme_minimal() +
        theme(panel.grid.major.y = element_blank(),
        panel.grid.minor.y = element_blank()))
```




```{r}
gwas_plot <- function (trait, max.pvalue) {
  
  # Colours and symbols I like using.
  clrs   <- rep(c("darkorange","dodgerblue","forestgreen","red",
                  "navyblue","gold","yellowgreen","darkviolet",
                  "magenta","cyan"),times = 4)
  shapes <- rep(c(20,1,2,4),each = 10)

  
   # Change the name of the "P-VALUE" column to "PVALUE".
  i <- which(colnames(gwas) == "P-VALUE")
  colnames(gwas)[i] <- "pValue"
  

  x <- subset(gwas,
              MAPPED_TRAIT == trait &
              as.numeric(pValue) < max.pvalue)
  
   x <- data.frame(
         effect.size = x[["OR or BETA"]],
         allele.freq = log10(as.numeric(x[["RISK ALLELE FREQUENCY"]])),
         study       = factor(x[["STUDY ACCESSION"]]),
         hla         = factor(grepl("HLA",x[["REPORTED GENE(S)"]]),
                              levels = c(FALSE,TRUE),labels = c("","HLA")))
  
    # The function's return value is a ggplot object. 
  return(ggplot(x,aes(x     = effect.size,
                      y     = allele.freq,
                      color = study,
                      shape = study,
                      label = hla)) +
    geom_point(size = 2.5) + 
    geom_text(size = 2.75,hjust = "left",vjust = "top",col = "black") + 
    scale_color_manual(values = clrs) +
    scale_shape_manual(values = shapes) +
    theme_minimal() +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank()) +
    labs(title = paste("GWAS Catalog associations for",trait,
                       "with p-value <",max.pvalue)))
}
```


```{R scatterplot_ms, warning=FALSE, eval=TRUE}
print(gwas_plot(trait = "malaria",max.pvalue = 1e-8))
```



### Significant malaria association SNPs from to GWAS catalog

```{r significant_snps_from_gwas_catalog}

signif_malaria_gwas <- subset(malaria_gwas, 
              as.numeric(pValue) <= 1e-8)
  
#View(signif_malaria_gwas)

  ## Select a subset of fields for display in the report  
  fields <- c(
    "rsID" = "SNPS",
    "Gene" = "MAPPED_GENE",
    "Phenotype" = "diseaseTrait")

    ## Display  the table
  signif_catalog_gwas <- signif_malaria_gwas[ , fields ]
  names(signif_catalog_gwas) <- names(fields)
   kable(signif_catalog_gwas)
   

```


## Data set of malaria genes and significant SNPs from to GWAS (catalog, 2014 and 2018)

```{r genes_snps}
## Merging the genes/SNPs

gene_snp_disease <- rbind(gwas.2014.vs.2018, signif_catalog_gwas)

#View(gene_snp_disease)

## Select TagSnps
TagSnps1 <- gene_snp_disease$rsID
TagSnps <- unique(str_replace(TagSnps1, pattern = "rs11335470_", replacement = "rs11335470"))

tagsnp.file <- "tagsnps.txt"
tagsnp.path <- file.path(dir.path["tagsnps"], tagsnp.file)
write.table(x = TagSnps,
              file = tagsnp.path,
              quote = FALSE,
              sep = "\t",
              row.names = FALSE,
              col.names = FALSE)

#View(TagSnps)
```





# Linkage Desequilibrium (LD) using HaploReg

```{r haplor}

## Run HaploReg with the tag SNPs
tagSnpsLdHaploreg <- queryHaploreg(query = TagSnps, ldThresh = 0.8, ldPop = "AFR", url = "http://archive.broadinstitute.org/mammals/haploreg/haploreg.php")

## Display a subset table from to haploreg
subsetLd <- tagSnpsLdHaploreg[c("GENCODE_name",  "chr", "pos_hg38", "rsID", "query_snp_rsid",  "ref", "alt", "r2", "Promoter_histone_marks", "Enhancer_histone_marks")]
kable(head(subsetLd))

## Export the full result table
 
 ld_haplo_file <- "tag_snps_haplo_ld.tsv"
 haplo_path <- file.path(dir.path["haploreg"], ld_haplo_file)
  message("\tLD table\t", haplo_path)

  write.table(x = tagSnpsLdHaploreg,
              file = haplo_path,
              quote = FALSE,
              sep = "\t",
              row.names = FALSE,
              col.names = TRUE)
  
#kable(head(tagSnpsLdHaploreg))
#View(tagSnpsLdHaploreg)
```

## Selection of SNPs Of Interest (SOIs)

```{r rsid_vs_ld_snps}
 LdTagsnps <- as.data.frame(unique(subsetLd$rsID, subsetLd$query_snp_rsid))

soi.file <- "snps_of_interest.txt"
soi.path <- file.path(dir.path["sois"], soi.file)
write.table(x = LdTagsnps,
              file = soi.path,
              quote = FALSE,
              sep = "\t",
              row.names = FALSE,
              col.names = FALSE)

```




