---
title: "Analysis of variants potentially associated with malaria"
author: 
  name: "Yvon Mbouamboua & Jacques van Helden"
affiliation: "Theories and Approach Genomic Complexity (TAGC)"
email: fridyvon@gmail.com
#date: 'Last update: `r Sys.Date()`'
date: Last update:`r format(Sys.time())`
output:
  html_document:
    code_folding: hide
    fig_caption: yes
    highlight: zenburn
    self_contained: yes
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: yes
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 3
  word_document: default
---

```{r libraries, include=FALSE,  eval=TRUE, echo=FALSE, warning=FALSE}
## Global configuration to generate the documents from this Rmd file
libraries.to.install <- c(
  "knitr",
  "readtext",
  "RCurl", 
  "ggplot2",
  "gridExtra", 
  "cowplot", 
  "dplyr",
  "tidyr",
  "VennDiagram", 
  "writexl", 
  "stringr", 
  "gdata", 
  "PythonInR", 
  "reticulate", 
  "haploR", 
  "RMySQL", 
  "GEOquery", 
  "gProfileR", 
  "lattice", 
  "qqman", 
  "readr", 
  "httr", 
  "jsonlite", 
  "xml2")

message("Loading libraries")
for (lib in libraries.to.install) {
  if (require(lib, character.only = TRUE, quietly = TRUE)) {
    message("\tLoaded library\t", lib)
  } else {
    message("Installing library\t", lib)
    install.packages(lib, dependencies = TRUE)
  }
}
```


```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo=FALSE, 
  eval=TRUE, 
  cache=TRUE, 
  message=FALSE, 
  warning=FALSE, 
  comment = "",  
  fig.align= "center",
  fig.width = 7, 
  fig.height = 5,
  fig.path = "figures/")

## Parameters
parameters <- list(
  update.flowcharts = TRUE, # Update the flowcharts with graphviz dot
  flowchart.formats = c("pdf", "png"), # List of formats to generate
  flowchart.format = "png" # Format for insertion in the report
)


## IDs of the diseases of interesd
diseaseIDs <- c("Malaria gene" = "C0024530", 
                "Malaria gene susceptibility" = "C1970028",
                "Antepartum malaria gene" = "C0747820",
                "Mild malaria gene susceptibility" = "C1836721",
                "Cerebral malaria gene" = "C0024534",
                "Cerebral malaria gene susceptibility" = "C1855457",
                "Cerebral malaria gene resistance" = "C1969379")

## Summarize the genes per disease from disgenet
gene.summary <- data.frame(
  row.names = diseaseIDs,
  ID = diseaseIDs,
  name = names(diseaseIDs))

## Get source
source("R/get_genes_for_disease.R")
source("R/get_snps_for_disease.R")
```



```{r configuration}
message("Defining working directory and subfolders")

## Main directory
dir.main <- '~/Google\ Drive/These_Yvon_2018/malaria_rSNPs'

# Result directory (export result tables)
dir.results <- file.path(dir.main, "results")
result.folders <- list(gene.disgenet = "gene.disgenet",
                       snp.disgenet = "snp.disgenet", 
                       gwas.2014 = "gwas.2014",
                       gwas.2018 = "gwas.2018",
                       remap = "remap",
                       roken = "roken",
                       tagsnps = "tagsnps",
                       haploreg = "haploreg",
                       sois = "sois",
                       soi.motifs = "soi.motifs")

dir.path <- c()
for (folder in result.folders) {
  dir.path[folder] <- file.path(dir.results, folder)
  dir.create(dir.path[folder], showWarnings = FALSE, recursive = TRUE)
}

message("\tWorking directory: ", getwd())

#setwd(dir.main)
```


# Goal

1) Identify the variants potentially associated with malaria.


# Bibliographic search

- Genes associated with malaria resistance.

- Look for rSNPs around these genes.

- Phenotypes studied:
  
  - Parasitemia
  - Cerebral malaria
  - Respiratory distress
  - Severe malaria
  - Severe malaria anemia


# GWAS search

- There are annotations of some symptoms in the GWAS catalog.
- To cross-search data of genes / SNPs associated with malaria.


# Search criteria for SNPs

- Number of studies / number of different populations where this variant was detected.
- Functional analyzes of the mutant.


## Flow chart 

```{r generate_flowcharts, eval=TRUE}

message("Updating flowchart figures")
for (flowchart in c(
  "flowchart/gene_wise_approach",
  "flowchart/snp_wise_approach",
  "flowchart/rsnp_pipline")) {
  
  for (format in parameters$flowchart.formats) {
    outfile <- paste(sep="", flowchart, ".", format)
    cmd <- paste(sep="", "dot -T", format, 
                 " ",  flowchart,".dot",
                 " -o ", outfile)
    #    message("\t", format, "\t", cmd)
    system(cmd)  
    message("\t", outfile)
  }
}
```

### Gene/SNP pipline

```{r fig.cap="Flow chart of the gene/SNP pipline. "}
knitr::include_graphics(paste(sep = "", "flowchart/rsnp_pipline.", parameters["flowchart.format"]))
```

### Gene-wise approach

```{r fig.cap="Flow chart of the gene-wise approach. "}
knitr::include_graphics(paste(sep = "", "flowchart/gene_wise_approach.", parameters["flowchart.format"]))
```

### SNP-wise approach

```{r fig.cap="Flow chart of the snp-wise approach. "}
knitr::include_graphics(paste(sep = "", "flowchart/snp_wise_approach.", parameters["flowchart.format"]))
```




# Collect the malaria candidate genes

## Data set of genes from to Huge Navigator (	https://phgkb.cdc.gov/PHGKB/hNHome.action). In particular, we use Phenotype (https://phgkb.cdc.gov/PHGKB/startPagePhenoPedia.action), subset of Huge Navigator, for to collect the genes associated with different malaria phenotypes.

```{r malaria_genes_phenopedia}
## Importing data from phenopedia web server
## We have collected the genes from to cerebral malaria, malaria P.falciparum, malaria P.vivax and malaria (phenotype non identified).
cerebral_malaria.pheno1 <- read_tsv('data/phenopedia/phenopedia_cerebral_malaria_genes.tsv')
malaria.pheno1 <- read_tsv('data/phenopedia/phenopedia_malaria_genes.tsv')
pf.pheno1 <- read_tsv('data/phenopedia/phenopedia_plasmodium_falciparum_genes.tsv')
pv.pheno1 <- read_tsv('data/phenopedia/phenopedia_plasmodium_vivax.tsv')

## Create new column for Disease
cerebral_malaria.pheno1 <- as.data.frame(cerebral_malaria.pheno1)
cerebral_malaria.pheno <-
  cerebral_malaria.pheno1 %>% 
  mutate(Disease = "Cerebral malaria")

malaria.pheno1 <- as.data.frame(malaria.pheno1)
malaria.pheno <-
  malaria.pheno1 %>% 
  mutate(Disease = "Malaria")

pf.pheno1 <- as.data.frame(pf.pheno1)
pf.pheno <-
  pf.pheno1 %>% 
  mutate(Disease = "Malaria P.falciparum")

pv.pheno1 <- as.data.frame(pv.pheno1)
pv.pheno <-
  pv.pheno1 %>% 
  mutate(Disease = "Malaria P.vivax")

## Combine all data
pheno.data <- rbind.data.frame(
  cerebral_malaria.pheno,
  malaria.pheno,
  pf.pheno, 
  pv.pheno)

print(ggplot(pheno.data, aes(AssociatedGene, fill = Disease)) + geom_bar() + coord_flip())
```





## Gene associated with malaria from DisGeNet

http://www.disgenet.org/web/DisGeNET/menu;jsessionid=1x6jf5ghyr1ga155fxkzfk2iji

```{r downloading_GDAs}

message("Downloading malaria-associated genes from disgenet")
## Download gene table from DisGeNet
url <- "http://www.disgenet.org/oql"

geneTables <- list()
genes.per.disease <- list()
gene.data.all.diseases <- data.frame()
for (d in 1:length(diseaseIDs)) {
  diseaseID <- diseaseIDs[[d]]
  diseaseName <- names(diseaseIDs)[[d]]
  
  gene.data <- GetGenesForDisease(diseaseID, diseaseName)
  gene.data.all.diseases <- rbind(gene.data.all.diseases, gene.data)
  
  ## Export the full result table
  gene.data.file <- paste(sep = "",
                          diseaseID, "_",
                          gsub(pattern = " +", replacement = "_", 
                               x = diseaseName),
                          ".tsv")
  gene.data.path <- file.path(dir.path["gene.disgenet"], gene.data.file)
  message("\tExporting gene table\t", gene.data.path)
  write.table(x = gene.data,
            file = gene.data.path,
            quote = FALSE,
            sep = "\t",
            row.names = FALSE,
            col.names = TRUE)
  
   ## Export the full result combined tables
  gene.data.file1 <- "all_genes_disease_association.tsv"
  gene.data.path1 <- file.path(dir.path["gene.disgenet"], gene.data.file1)
  message("\tExporting gene table\t", gene.data.path1)
  write.table(x = gene.data.all.diseases,
            file = gene.data.path1,
            quote = FALSE,
            sep = "\t",
            row.names = FALSE,
            col.names = TRUE)

}



# venn.file <- file.path(dir.results, "genes.per.disease.Venn.tiff")
# venn.diagram(x = genes.per.disease, filename = venn.file) #, height = 7, width = 8, resolution = 72)
# venn.plot <- venn.diagram(x = genes.per.disease, filename = venn.file) #, height = 7, width = 8, resolution = 72)

```

## SNPs associated with malaria from DisGeNet





```{r}
#print(ggplot(snp.data, aes(c2.most_severe_consequence	)) + geom_bar() + coord_flip())
```




## Collecte the significant SNPs associated with malaria from to GWAS 2014 and 2018

```{r gwas_2014_2018}
gwas2014 <- read_tsv("data/gwas/mgen_2014_signif_SNPs.tsv")
gwas2018 <- read_tsv("data/gwas/ravhenall_2018_signif_SNPs.tsv")

## Select a subset of fields for display in the report
snp.gwas.fields <- c("Gene", "rsID", "Phenotype") 
  
## Display a small subset of the table
gwas2014.table <- head(gwas2014[, snp.gwas.fields])

gwas2018.table <- head(gwas2018[, snp.gwas.fields])

   
## Combine datset from GWAS 2014 and 2018
gwas_2014_vs_2018 <- rbind(gwas2014.table, gwas2018.table)
#View(gwas_2014_vs_2018)
kable(gwas_2014_vs_2018)

## Plot the interest genes
print(ggplot(gwas_2014_vs_2018, aes(Phenotype)) + geom_bar() + coord_flip())
print(ggplot(gwas_2014_vs_2018, aes(Gene)) + geom_bar() + coord_flip())
```




