---
title: "Analysis of variants potentially associated with malaria"
author: 
  name: "Yvon Mbouamboua & Jacques van Helden"
affiliation: "Theories and Approach Genomic Complexity (TAGC)"
email: fridyvon@gmail.com
#date: 'Last update: `r Sys.Date()`'
date: Last update:`r format(Sys.time())`
output:
  html_document:
    code_folding: hide
    fig_caption: yes
    highlight: zenburn
    self_contained: yes
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: yes
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 3
  word_document: default
---

```{r libraries, include=FALSE,  eval=TRUE, echo=FALSE, warning=FALSE}
## Global configuration to generate the documents from this Rmd file
libraries.to.install <- c(
  "knitr",
  "readtext",
  "RCurl", 
  "ggplot2",
  "gridExtra", 
  "cowplot", 
  "dplyr",
  "tidyr",
  "VennDiagram", 
  "writexl", 
  "stringr", 
  "gdata", 
  "PythonInR", 
  "reticulate", 
  "haploR", 
  "RMySQL", 
  "GEOquery", 
  "gProfileR", 
  "lattice", 
  "qqman", 
  "readr", 
  "httr", 
  "jsonlite", 
  "xml2")

message("Loading libraries")
for (lib in libraries.to.install) {
  if (require(lib, character.only = TRUE, quietly = TRUE)) {
    message("\tLoaded library\t", lib)
  } else {
    message("Installing library\t", lib)
    install.packages(lib, dependencies = TRUE)
  }
}
```


```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo=FALSE, 
  eval=TRUE, 
  cache=TRUE, 
  message=FALSE, 
  warning=FALSE, 
  comment = "",  
  fig.align= "center",
  fig.width = 7, 
  fig.height = 5,
  fig.path = "figures/")

## Parameters
parameters <- list(
  update.flowcharts = TRUE, # Update the flowcharts with graphviz dot
  flowchart.formats = c("pdf", "png"), # List of formats to generate
  flowchart.format = "png" # Format for insertion in the report
)


## IDs of the diseases of interesd according to the genes
diseaseIDs <- c("Malaria gene" = "C0024530", 
                "P.falciparum Malaria gene" = "C0024535",
                "P.vivax Malaria gene" = "C0024537",
                "Malaria gene susceptibility" = "C1970028",
                "Antepartum malaria gene" = "C0747820",
                "Mild malaria gene susceptibility" = "C1836721",
                "Cerebral malaria gene" = "C0024534",
                "Cerebral malaria gene susceptibility" = "C1855457",
                "Cerebral malaria gene resistance" = "C1969379")

## Summarize the genes per disease from disgenet
disease.summary <- data.frame(
  row.names = diseaseIDs,
  ID = diseaseIDs,
  name = names(diseaseIDs))
disease.summary <- rbind(disease.summary, 
                         data.frame(
                           row.names = "all",
                           ID = "all", 
                           name = "All selected diseases"))
#kable(disease.summary, row.names = FALSE)
                         

# 
# ## Summarize the genes per disease from disgenet
# snp.summary <- data.frame(
#   row.names = diseaseIDs,
#   ID = diseaseIDs,
#   name = names(diseaseIDs))
# #kable(snp.summary, row.names = FALSE)

```



```{r configuration}
message("Defining working directory and subfolders")

## Main directory
dir.main <- '~/Google\ Drive/These_Yvon_2018/malaria_rSNPs'

# Result directory (export result tables)
dir.results <- file.path(dir.main, "results")
result.folders <- list(gene_disgenet = "disgenet_genes",
                       snp_disgenet = "disgenet_snps", 
                       gwas_2014 = "gwas_2014",
                       gwas_2018 = "gwas_2018",
                       remap = "remap",
                       roken = "roken",
                       tagsnps = "tagsnps",
                       ldtagsnp = "ldtagsnp",
                       haploreg = "haploreg",
                       sois = "sois",
                       soi.motifs = "soi.motifs")

dir.path <- c()
for (folder in result.folders) {
  dir.path[folder] <- file.path(dir.results, folder)
  dir.create(dir.path[folder], showWarnings = FALSE, recursive = TRUE)
}

message("\tWorking directory: ", getwd())

#setwd(dir.main)
```


# Goal

1) Identify the variants potentially associated with malaria.


# Bibliographic search

- Genes associated with malaria resistance.

- Look for rSNPs around these genes.

- Phenotypes studied:
  
  - Parasitemia
  - Cerebral malaria
  - Respiratory distress
  - Severe malaria
  - Severe malaria anemia


# GWAS search

- There are annotations of some symptoms in the GWAS catalog.
- To cross-search data of genes / SNPs associated with malaria.


# Search criteria for SNPs

- Number of studies / number of different populations where this variant was detected.
- Functional analyzes of the SNP.


# Flow chart 

```{r generate_flowcharts, eval=TRUE}

message("Updating flowchart figures")
for (flowchart in c(
  "flowchart/gene_wise_approach",
  "flowchart/snp_wise_approach",
  "flowchart/rsnp_pipeline")) {
  
  for (format in parameters$flowchart.formats) {
    outfile <- paste(sep="", flowchart, ".", format)
    cmd <- paste(sep="", "dot -T", format, 
                 " ",  flowchart,".dot",
                 " -o ", outfile)
    #    message("\t", format, "\t", cmd)
    system(cmd)  
    message("\t", outfile)
  }
}
```

### Gene/SNP pipeline

```{r fig.cap="Flow chart of the gene/SNP pipeline. "}
knitr::include_graphics(paste(sep = "", "flowchart/rsnp_pipeline.", parameters["flowchart.format"]))
```

### Gene-wise approach

```{r fig.cap="Flow chart of the gene-wise approach. "}
knitr::include_graphics(paste(sep = "", "flowchart/gene_wise_approach.", parameters["flowchart.format"]))
```

### SNP-wise approach

```{r fig.cap="Flow chart of the snp-wise approach. "}
knitr::include_graphics(paste(sep = "", "flowchart/snp_wise_approach.", parameters["flowchart.format"]))
```




# Collect the malaria candidate genes

## Data set of genes from to Huge Navigator 

Web server of Huge Navigator:https://phgkb.cdc.gov/PHGKB/hNHome.action.
In particular, we use Phenotype (https://phgkb.cdc.gov/PHGKB/startPagePhenoPedia.action), subset of Huge Navigator, for to collect the genes associated with different malaria phenotypes.

```{r malaria_genes_phenopedia}
## Importing data from phenopedia web server
## We have collected the genes from to cerebral malaria, malaria P.falciparum, malaria P.vivax and malaria (phenotype non identified).
cerebral_malaria.pheno1 <- read_tsv('data/phenopedia/phenopedia_cerebral_malaria_genes.tsv')
malaria.pheno1 <- read_tsv('data/phenopedia/phenopedia_malaria_genes.tsv')
pf.pheno1 <- read_tsv('data/phenopedia/phenopedia_plasmodium_falciparum_genes.tsv')
pv.pheno1 <- read_tsv('data/phenopedia/phenopedia_plasmodium_vivax.tsv')

## Create new column for Disease
cerebral_malaria.pheno1 <- as.data.frame(cerebral_malaria.pheno1)
cerebral_malaria.pheno <-
  cerebral_malaria.pheno1 %>% 
  mutate(Disease = "Cerebral malaria")

malaria.pheno1 <- as.data.frame(malaria.pheno1)
malaria.pheno <-
  malaria.pheno1 %>% 
  mutate(Disease = "Malaria")

pf.pheno1 <- as.data.frame(pf.pheno1)
pf.pheno <-
  pf.pheno1 %>% 
  mutate(Disease = "Malaria P.falciparum")

pv.pheno1 <- as.data.frame(pv.pheno1)
pv.pheno <-
  pv.pheno1 %>% 
  mutate(Disease = "Malaria P.vivax")

## Combine all data
pheno.data <- rbind.data.frame(
  cerebral_malaria.pheno,
  malaria.pheno,
  pf.pheno, 
  pv.pheno)

kable(head(pheno.data), caption = "Genes associated malaria from to Phenopedia.")
#print(ggplot(pheno.data, aes(AssociatedGene, fill = Disease)) + geom_bar() + coord_flip())
```

**We find `r nrow(pheno.data)` genes associated with malaria from to Phenopedia.**


## Gene associated with malaria from DisGeNet

http://www.disgenet.org/web/DisGeNET/menu;jsessionid=1x6jf5ghyr1ga155fxkzfk2iji

```{r downloading_GDAs}

## Get source
source("R/get_genes_for_disease.R")

message("Downloading malaria-associated genes from disgenet")
## Download gene table from DisGeNet
url <- "http://www.disgenet.org/oql"

geneTables <- list()
genes.per.disease <- list()
gene.data.all.diseases <- data.frame()
for (d in 1:length(diseaseIDs)) {
  diseaseID <- diseaseIDs[[d]]
  diseaseName <- names(diseaseIDs)[[d]]
  
  gene.data <- GetGenesForDisease(diseaseID, diseaseName)
  gene.data.all.diseases <- rbind(gene.data.all.diseases, gene.data)
  
  ## Export the disease-specific result table
  gene.data.file <- paste(sep = "",
                          diseaseID, "_",
                          gsub(pattern = " +", replacement = "_", 
                               x = diseaseName),
                          ".tsv")
  gene.data.path <- file.path(dir.path["disgenet_genes"], gene.data.file)
  
  message("\tExporting gene table\t", gene.data.path)
  write.table(x = gene.data,
            file = gene.data.path,
            quote = FALSE,
            sep = "\t",
            row.names = FALSE,
            col.names = TRUE)
  
  ## Add information to the gene summary table
  disease.summary[diseaseID, "disgenet_genes"] <- nrow(gene.data)
  disease.summary[diseaseID, "disgenet_gene_table"] <- gene.data.path
  

#   venn.file <- file.path(dir.results, "genes.per.disease.Venn.tiff")
# venn.diagram(x = genes.per.disease, filename = venn.file) #, height = 7, width = 8, resolution = 72)
# venn.plot <- venn.diagram(x = genes.per.disease, filename = venn.file) #, height = 7, width = 8, resolution = 72)
# print(venn.plot)
}

## Export the full result combined tables
gene.data.file <- "all_genes_disease_association.tsv"
gene.data.path <- file.path(dir.path["disgenet_genest"], gene.data.file)
message("\tExporting gene table\t", gene.data.path)
write.table(x = gene.data.all.diseases,
            file = gene.data.path,
            quote = FALSE,
            sep = "\t",
            row.names = FALSE,
            col.names = TRUE)

## Add information to the gene summary table
disease.summary["all", "disgenet_genes"] <- nrow(gene.data.all.diseases)
disease.summary["all", "disgenet_gene_table"] <- gene.data.path

# View(disease.summary)

```

**DisGeNet provides `r nrow(gene.data.all.diseases)` genes associated with different malaria phenotypes: [`r gene.data.file`](`r gene.data.path`)**


## SNPs associated with malaria from DisGeNet

```{r downloading VDAs}
## Get source
source("R/get_snps_for_disease.R")

message("Downloading malaria-associated SNPs from disgenet")
## Download SNP table from DisGeNet
url <- "http://www.disgenet.org/oql"

snpTables <- list()
snp.per.disease <- list()
snp.data.all.diseases <- data.frame()
for (s in 1:length(diseaseSnpIDs)) {
  diseaseSnpID <- diseaseSnpIDs[[s]]
  diseaseName <- names(diseaseSnpIDs)[[s]]
  
  snp.data <- GetSnpsForDisease(diseaseSnpID, diseaseName)
  snp.data.all.diseases <- rbind(snp.data.all.diseases, snp.data)
  
  ## Export the full result table
  snp.data.file <- paste(sep = "",
                          diseaseSnpID, "_",
                          gsub(pattern = " +", replacement = "_", 
                               x = diseaseName),
                          ".tsv")
  
  snp.data.path <- file.path(dir.path["disgenet_snpst"], snp.data.file)
  message("\tgene table\t", snp.data.path)

  write.table(x = snp.data,
              file = snp.data.path,
              quote = FALSE,
              sep = "\t",
              row.names = FALSE,
              col.names = TRUE)
  
    ## Export the full result combined tables
  snp.data.file1 <- "all_snps_disease_association.tsv"
  snp.data.path1 <- file.path(dir.path["disgenet_snpst"], snp.data.file1)
  message("\tExporting snp table\t", snp.data.path1)
  write.table(x = snp.data.all.diseases,
            file = snp.data.path1,
            quote = FALSE,
            sep = "\t",
            row.names = FALSE,
            col.names = TRUE)
  
}
```



```{r ggplot_functional_annotation_and_phenotypes}

print(ggplot(snp.data.all.diseases, 
             aes(c2.most_severe_consequence	)) + 
             geom_bar() + 
             coord_flip() +
             ggtitle("Different malaria functional SNP annotations") +
             xlab("Functional annotation") +
             ylab("Number of SNPs"))


print(ggplot(snp.data.all.diseases, 
             aes(c1.name)) + 
             geom_bar() + 
             coord_flip() +
             ggtitle("Different malaria phenotypes") +
             xlab("Phenotypes") +
             ylab("Number of SNPs"))
```


```{r snps_vs_phenotypes}
snp.fields <- c(
  "chr" = "c2.chromosome",
  "rsid" = "c0.snpId",
  "ref_alt" = "c2.REF_ALT",
  "consequence" = "c2.most_severe_consequence",
  "npmid" = "c0.Npmids",
  "phenotype" = "c1.name"
)

## Display a small subset of the table
  snp.table <- snp.data.all.diseases[, snp.fields]
  names(snp.table) <- names(snp.fields)
   kable(head(snp.table), caption = "SNPs associated with different malaria phenotypes from DisGeNet.")
```

**We are collected `r nrow(snp.data.all.diseases)` SNPs associated with different malaria phenotypes from to DisGeNet: [`r snp.data.file1`](`r snp.data.path1`)**


### Select all studied SNPs more than one population from DisGeNet

Only SNPs that had a dbSNP identifier and whose trait association was replicated in at least two studies were considered in this analyse. 

```{r snps_vs_populations}
all_snps_disgenet <- read_tsv(snp.data.path1)
snp2populations <-
  all_snps_disgenet %>%
  filter(c0.Npmids > 1)
#View(snp2populations)
```

**We are a subset of `r nrow(snp2populations)` SNPs identified in more than one studies.**


## Collecte the significant Genes and SNPs associated with malaria from to GWAS 2014 and 2018

```{r gwas_2014_2018}
## Importing data from GWAS 2014 and 2018
gwas2014 <- read_tsv("data/gwas/mgen_2014_signif_SNPs.tsv")
gwas2018 <- read_tsv("data/gwas/ravhenall_2018_signif_SNPs.tsv")

## Tranformation the data in data.frame
gwas2014 <- as.data.frame(gwas2014)
gwas2018 <- as.data.frame(gwas2018)

## Select a subset of fields for display in the report
gwas_gene_snp2014 <-  gwas2014[, c("gene", "rsid", "phenotype")]
gwas_gene_snp2018 <- gwas2018[, c("gene", "rsid", "phenotype")]

## Combine datset from GWAS 2014 and 2018
gwas_2014_vs_2018 <- rbind(gwas_gene_snp2014, gwas_gene_snp2018)
#View(gwas_2014_vs_2018)
kable(head(gwas_2014_vs_2018))

## Plot the interest genes
#print(ggplot(gwas_2014_vs_2018, aes(phenotype)) + geom_bar() + coord_flip())
print(ggplot(gwas_2014_vs_2018, aes(gene)) + 
        geom_bar() + 
        coord_flip() +
        ggtitle("Genes of interest from to malaria GWAS (MGEN, 2014 and Ravhenall                  et al, 2018") +
        xlab("Genes") +
        ylab("Number of genes"))
```

**The malaria GWAS from to different country and African population provide `r nrow(gwas_2014_vs_2018)` significant SNPs.**

### Tag SNPs

The Tag SNPs are the result of the combinaison of the SNPs from DisGeNet and GWAS (2014, 2018)

```{r disgenet_vs_gwas_2014_2018}
## TagSNP-phenotype from to GWAS (2014 and 2018)
GwasTagSnp <- gwas_2014_vs_2018[, c("rsid", "phenotype")]

## TagSNP-phenotype from to disgenet
DisgenetTagSnp <- as.data.frame(snp2populations[, c("c0.snpId", "c1.name")])

## Rename the columns of DisgenetTagSnp variable
names(DisgenetTagSnp) <- c("rsid", "phenotype")

#View(DisgenetTagSnp)

Disgenet_vs_gwas <- rbind(DisgenetTagSnp, GwasTagSnp)
TagSnps <- unique(Disgenet_vs_gwas$rsid)

tagsnps.file <- "tagsnps.txt"
tagsnps.path <- file.path(dir.path["tagsnps"], tagsnps.file)
write.table(x = TagSnps,
              file = tagsnps.path,
              quote = FALSE,
              sep = "\t",
              row.names = FALSE,
              col.names = FALSE)
```


**We collected a set of the `r length (TagSnps)` Tag SNPs associated with different malaria phenotypes: [`r tagsnps.file`](`r tagsnps.path`)**



# Haplotype collection

We used the R HalpoR package which allows to recover SNPs in linkage Desequilibrium (LD) using HaploReg by specifying the population. We specified the African population and submitted the SNPs of interest obtained by the 3 requests above (DisGenet GWAS malaria 2014 and GWAS malaria 2018).


```{r haplor}

## Run HaploReg with the tag SNPs
tagSnpsLdHaploreg <- queryHaploreg(query = TagSnps, ldThresh = 0.8, ldPop = "AFR", url = "http://archive.broadinstitute.org/mammals/haploreg/haploreg.php")

## Export the full result table
 
 ld_haplo_file <- "tag_snps_haplo_ld.tsv"
 haplo_path <- file.path(dir.path["haploreg"], ld_haplo_file)
  message("\tLD table\t", haplo_path)

  write.table(x = tagSnpsLdHaploreg,
              file = haplo_path,
              quote = FALSE,
              sep = "\t",
              row.names = FALSE,
              col.names = TRUE)
  
#kable(head(tagSnpsLdHaploreg))
#View(tagSnpsLdHaploreg)
  
## Display a subset table from to haploreg
subsetLd <- tagSnpsLdHaploreg[c("chr", 
                                "GENCODE_name", 
                                "rsID", 
                                "ref", 
                                "alt", 
                                "query_snp_rsid", 
                                "r2", 
                                "AFR"
                                )]
kable(head(subsetLd))
```

**We find `r length(unique(subsetLd$rsID))` LD SNPs with `r length (TagSnps)` Tag SNPs**


## SNPs of interest (SOIs)

The SOIs is the combinaison of the TagSNPs and the LD SNPs.

```{r rsid_vs_ld_snps}

LdTagsnps1 <- unique(tagSnpsLdHaploreg$rsID, tagSnpsLdHaploreg$query_snp_rsid)
LdTagsnps <- as.data.frame(unique(LdTagsnps1))
names(LdTagsnps) <- "rsid"
LdTagsnps <- as.data.frame(LdTagsnps)

#kable(head(LdTagsnps), caption = "Liste of SNPs of interest (SOI)", col.names = "rsID")

## In our list of SOIs, these SNPs are merged in rs55868763 = rs1541254
## rs79775818 = rs1541253, rs112685218 = rs8176733

 LdTagsnps.data <-
   LdTagsnps%>%
   mutate(
     rsid = as.character(rsid),
     snp1 = str_detect(rsid, "rs55868763"),
          rsid = ifelse(snp1 == TRUE, "rs1541254", rsid),
     snp2 = str_detect(rsid, "rs79775818"),
          rsid = ifelse(snp2 == TRUE, "rs1541253", rsid),
     snp3 = str_detect(rsid, "rs112685218"),
          rsid = ifelse(snp3 == TRUE, "rs8176733", rsid))
    
 #View(LdTagsnps.data)

## Export the full result table
ld.file <- "ldtagsnp.txt"
ld.path <- file.path(dir.path["ldtagsnp"], ld.file)
write.table(x = LdTagsnps,
              file = ld.path,
              quote = FALSE,
              sep = "\t",
              row.names = FALSE,
              col.names = FALSE)
```

**We are a total `r length(LdTagsnps1)` [`r ld.file`](`r ld.path`) SNPs of interest (SOIs)**



## Regulons of Interest (ROIs)

 We define the ROIs for the SNPs From the SOIs using the intersection with functional enhancers and promoters (from to HaploReg).
 
```{r soi}
## Selection of the ROI according to functional enhancers and promoters and the non-coding region.
roi <- subset(tagSnpsLdHaploreg, 
              Enhancer_histone_marks != ""
              & Promoter_histone_marks != ""
              & dbSNP_functional_annotation == "INT")

## Plot of ROIs
print(ggplot(roi, 
             aes(GENCODE_name	)) + 
             geom_bar() + 
             coord_flip() +
             ggtitle("Malaria genes as a function of SNPs withim promoters and                         enhancers histone marks and non-coding region") +
             xlab("Genes") +
             ylab("Count"))
#View(roi)

## Selection of the ROI according to functional enhancers and promoters, the non-coding region and the eQTL.
roi1 <- subset(tagSnpsLdHaploreg, 
              Enhancer_histone_marks != ""
              & Promoter_histone_marks != ""
              & dbSNP_functional_annotation == "INT"
              & eQTL != "" & eQTL != ".")

print(ggplot(roi1, 
             aes(GENCODE_name	)) + 
             geom_bar() + 
             coord_flip() +
             ggtitle("Malaria genes as a function of SNPs withim promoters and                         enhancers histone marks, non-coding region and eQTL") +
             xlab("Genes") +
             ylab("Count"))

## Display a subset LD SNPs in enhancer histone marks, promoters histone marks and non-condant region.

ld.fields <- tagSnpsLdHaploreg[, c("chr",
                                   "GENCODE_name",
                                   "pos_hg38",
                                   "r2",
                                   "rsID",
                                   "query_snp_rsid",
                                   "ref",
                                   "alt",
                                   "AFR",
                                   "dbSNP_functional_annotation",
                                   "Promoter_histone_marks",
                                   "Enhancer_histone_marks")]

kable(head(ld.fields), caption = "Table of Regulons Of Interest (ROIs)")
```

**According to the SNPs finded in non-conding region, enhancers and promoters histone marks, we selected a subset of `r nrow(roi)` Reguloms of interest (ROIs).**

**If we considere also the eQTLs, the ROIs provide `r nrow(roi1)` SNPs.**


# Analyse of Genes/SNPs from GWAS malaria

```{r gwas_malaria}
## Importing gwas malaria data
gwasMalaria <- read.xls('data/gwasMalaria.xlsx', sheet = 1)

## Barplot of malaria genes associted

print(ggplot(gwasMalaria, 
             aes(gene	)) + 
             geom_bar() + 
             coord_flip() +
             ggtitle("Malaria genes from to GWAS") +
             xlab("Genes") +
             ylab("Count"))


## Barplot of malaria phenotype associted

print(ggplot(gwasMalaria, 
             aes(diseaseTrait)) + 
             geom_bar() + 
             coord_flip() +
             ggtitle("Malaria phenotype from to GWAS") +
             xlab("Phenotype") +
             ylab("Count"))

## Selection LD SNPs using haploreg
### Selection of the TagSNPs
TagSNPs <- gwasMalaria$rsId

### Run haploreg with the TagSNPS
TagSNPs.ld <- queryHaploreg(query = TagSNPs, 
                            ldThresh = 0.8, 
                            ldPop = "AFR", 
                            url = "http://archive.broadinstitute.org/mammals/haploreg/haploreg.php")

## Selection of the ROI according to functional enhancers and promoters, the non-coding region and the eQTL.
ROIs <- subset(TagSNPs.ld, 
              Enhancer_histone_marks != ""
              & Promoter_histone_marks != ""
              & dbSNP_functional_annotation == "INT"
              & eQTL != "" & eQTL != ".")


```




