---
title: "Analysis of variants potentially associated with malaria"
author: 
  name: "Yvon Mbouamboua & Jacques van Helden"
affiliation: "Theories and Approach Genomic Complexity (TAGC)"
email: fridyvon@gmail.com
#date: 'Last update: `r Sys.Date()`'
date: Last update:`r format(Sys.time())`
output:
  html_document:
    code_folding: hide
    fig_caption: yes
    highlight: zenburn
    self_contained: yes
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: yes
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 3
  word_document: default
---

```{r libraries, include=FALSE,  eval=TRUE, echo=FALSE, warning=FALSE}
## Global configuration to generate the documents from this Rmd file
libraries.to.install <- c(
  "knitr",
  "readtext",
  "RCurl", 
  "ggplot2",
  "gridExtra", 
  "cowplot", 
  "dplyr",
  "tidyr",
  "VennDiagram", 
  "writexl", 
  "stringr", 
  "gdata", 
  "PythonInR", 
  "reticulate", 
  "haploR", 
  "RMySQL", 
  "GEOquery", 
  "gProfileR", 
  "lattice", 
  "qqman", 
  "readr", 
  "httr", 
  "jsonlite", 
  "xml2")

message("Loading libraries")
for (lib in libraries.to.install) {
  if (require(lib, character.only = TRUE, quietly = TRUE)) {
    message("\tLoaded library\t", lib)
  } else {
    message("Installing library\t", lib)
    install.packages(lib, dependencies = TRUE)
  }
}
```


```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo=FALSE, 
  eval=TRUE, 
  cache=TRUE, 
  message=FALSE, 
  warning=FALSE, 
  comment = "",  
  fig.align= "center",
  fig.width = 7, 
  fig.height = 5,
  fig.path = "figures/")

## Parameters
parameters <- list(
  update.flowcharts = TRUE, # Update the flowcharts with graphviz dot
  flowchart.formats = c("pdf", "png"), # List of formats to generate
  flowchart.format = "png" # Format for insertion in the report
)


## IDs of the diseases of interesd
diseaseIDs <- c("Malaria" = "C0024530", 
                "Malaria susceptibility" = "C1970028",
                "Antepartum malaria" = "C0747820",
                "Mild malaria susceptibility" = "C1836721",
                "Cerebral malaria" = "C0024534",
                "Cerebral malaria susceptibility" = "C1855457",
                "Cerebral malaria resistance" = "C1969379")

## Summarize the genes per disease from disgenet
gene.summary <- data.frame(
  row.names = diseaseIDs,
  ID = diseaseIDs,
  name = names(diseaseIDs))


```



```{r configuration}
## Main directory
dir.main <- '~/Google\ Drive/These_Yvon_2018/malaria_rSNPs'

# Result directory (export result tables)
dir.results <- file.path(dir.main, "results")
result.folders <- list(disgenet = "disgenet",
                       gwas.2014 = "gwas.2014",
                       gwas.2018 = "gwas.2018",
                       remap ="remap",
                       roken ="roken",
                       disgenet_vs_gwas = "disgenet_vs_gwas",
                       TagSNPs = "TagSNPs",
                       bedtools ="bedtools", 
                       haploreg ="haploreg",
                       phenotype = "phenotype",
                       ucsc = "ucsc",
                       SOIs = "SOIs",
                       SOI_motifs = "SOI_motifs")

dir.path <- c()
for (folder in result.folders) {
  dir.path[folder] <- file.path(dir.results, folder)
  dir.create(dir.path[folder], showWarnings = FALSE, recursive = TRUE)
}

message("Working directory: ", getwd())

#setwd(dir.main)
```


# Goal

1) Identify the variants potentially associated with malaria.


# Bibliographic search

- Genes associated with malaria resistance.

- Look for rSNPs around these genes.

- Phenotypes studied:
  
  - Parasitemia
- Cerebral malaria
- Respiratory distress
- Severe anemia


# GWAS search

- There are annotations of some symptoms in the GWAS catalog.
- To cross-search data of genes / SNPs associated with malaria.


# Search criteria for SNPs

- Number of studies / number of different populations where this variant was detected.
- Functional analyzes of the mutant.


# Flow chart 

```{r generate_flowcharts, eval=TRUE}

message("Updating flowchart figures")
for (flowchart in c(
  "flowchart/gene_wise_approach",
  "flowchart/snp_wise_approach")) {
  
  for (format in parameters$flowchart.formats) {
    outfile <- paste(sep="", flowchart, ".", format)
    cmd <- paste(sep="", "dot -T", format, 
                 " ",  flowchart,".dot",
                 " -o ", outfile)
    #    message("\t", format, "\t", cmd)
    system(cmd)  
    message("\t", outfile)
  }
}

```

## Gene-wise approach

```{r fig.cap="Flow chart of the gene-wise approach. "}
knitr::include_graphics(paste(sep="", "flowchart/gene_wise_approach.", parameters["flowchart.format"]))
```

## SNP-wise approach

```{r fig.cap="Flow chart of the snp-wise approach. "}
knitr::include_graphics(paste(sep="", "flowchart/snp_wise_approach.", parameters["flowchart.format"]))
```


# Tools of to collect gÃ¨nes/SNPs cancidats selection

1. DisGenet
http://www.disgenet.org/web/DisGeNET/menu;jsessionid=1x6jf5ghyr1ga155fxkzfk2iji

## Gene associated with malaria from DisGeNet


```{r downloading_GDAs}

message("Downloading malaria-associated genes from disgenet")
## Download gene table from DisGeNet
url <- "http://www.disgenet.org/oql"

GetGenesForDisease <- function (diseaseID) {
  message("\tGetGenesForDisease\t", diseaseID)
  
  ## Generate the SQL query for DisGeNet
  oql <- paste(sep="", "DEFINE
               c0='/data/gene_disease',
               c1='/data/genes',
               c2='/data/diseases',
               c3='/data/gene_disease_summary',
               c4='/data/publication',
               c5='/data/sources'
               ON
               'http://www.disgenet.org/web/DisGeNET'
               SELECT
               c0 (source, geneId, source, geneId, source, geneId, associationType,                         originalSource, originalSource, originalSource, sentence, pmid),
               c1 (pantherName, symbol, geneId, description, symbol, geneId),
               c2 (diseaseId, name, hpoName, diseaseId, name, STY, MESH,                                    diseaseClassName, doName, type, OMIM, type),
               c3 (score),
               c4 (year)
               FROM
               c0
               WHERE
               (
               c2 = '", diseaseID, "'
               AND
               c5 = 'ALL'
               )
               ORDER BY
               c3.score DESC" )
  
  
  ## Open a connection to disgenet and get the full table of disease-associated SNPs
  dataTsv <- rawToChar(
    charToRaw( 
      getURLContent(url, 
                    readfunction = charToRaw(oql), 
                    upload = TRUE, 
                    customrequest = "POST")))
  gene.data <- read.csv(textConnection(dataTsv), header = TRUE, sep="\t")
  

  # dim(gene.data)
  #View(gene.data)
  message("\t", nrow(gene.data), "\tgenes associated to disease ", diseaseID)
  
  
  
   geneTables <- list()
d <- 2
genes.per.disease <- list()
for (d in 1:length(diseaseIDs)) {
  diseaseID <- diseaseIDs[[d]]
  diseaseName <- names(diseaseIDs)[[d]]


  geneTables[[diseaseID]] <- GetGenesForDisease(diseaseID)
  gene.summary[diseaseID, "nb_genes"] <- nrow(geneTables[[diseaseID]])

  gene.symbols <- as.vector(unlist(geneTables[[diseaseID]]$c1.symbol))

  if (length(gene.symbols) > 0) {
    genes.per.disease[[diseaseName]] <-gene.symbols
  }

}


  return(gene.data)   


}


## Export the full result table
  gene.data.file <- paste(sep="", 
                          diseaseID, "_", 
                          gsub(pattern = " +", replacement = "_", x = diseaseName),
                          ".tsv")
  
  gene.data.path <- file.path(dir.path["disgenet"], gene.data.file)
  message("\tgene table\t", gene.data.path)
  write.table(x = gene.data,
              file = gene.data.file,
              quote = FALSE,
              sep = "\t",
              row.names = FALSE,
              col.names = TRUE)

  ## Select a subset of fields for display in the report
  gene.fields <- c(
    "Gene" = "c1.symbol",
    "Association type" = "c0.associationType",
    "Score" = "c3.score",
    "Disease" = "c2.name",
    PMID = "c0.pmid"
  ) 
     
  ## Display a small subset of the table
  display.table <- head(gene.data[, gene.fields])
  names(display.table) <- names(gene.fields)
   kable(display.table)

 
  
 
venn.file <- file.path(dir.results, "genes_per_disease_Venn.tiff")
venn.diagram(x = genes.per.disease, filename = venn.file) #, height = 7, width = 8, resolution = 72)
venn.plot <- venn.diagram(x = genes.per.disease, filename = venn.file) #, height = 7, width = 8, resolution = 72)




```





